### ISSUE TITLE
Infill offset when adaptive slicing layer height changes

### ISSUE THREAD

**Issue by mat813** [slic3r/Slic3r#4705]
**Describe the bug**
When the layer height changes abruptly, the infill layers shifts, sometime ending up over empty air, fail to have any support at all.

**To Reproduce**
Use the provided model and config, render it.  Infill shifts between layers heights 38.31 and 38.96, and then 49.94 and 51.59. And it starts when the layer height changes, then stabilizes.

**3D Model and Slic3r Configuration Export**
[Archive.zip](https://github.com/slic3r/Slic3r/files/2761771/Archive.zip)

**Expected behavior**
The infill layers should stay aligned.

**Screenshots**
<img width="672" alt="slic3r" src="https://user-images.githubusercontent.com/75783/51214772-2e1f6200-191f-11e9-9a40-c4b4338390d6.png">
<img width="677" alt="slic3r2" src="https://user-images.githubusercontent.com/75783/51214773-2eb7f880-191f-11e9-8e55-a33270aa8f1b.png">
<img width="677" alt="slic3r" src="https://user-images.githubusercontent.com/75783/51214885-80f91980-191f-11e9-8dd8-f5e7c641ffa7.png">
<img width="684" alt="slic3r2" src="https://user-images.githubusercontent.com/75783/51214887-80f91980-191f-11e9-88e4-30c30bfe9912.png">

**Desktop (please complete the following information):**
 - OS: MacOS
 - Version [e.g. 1.3.0] 916c30c

**Additional context**
I exported the config using the "stars" infill pattern, and added screenshots with stars and gyroid, the shift happens on all infill patterns.

**Comment by Lucki** [comment]
I've also noticed this with honeycomb. The infill shifts on every adaptive layer height changes.

Desktop:

* OS: Arch Linux
* Version: 1.3.0

**Comment by lordofhyphens** [comment]
Adaptive is probably breaking one of assumptions in that codebase, that or the line thickness is changing (which would rescale the polygon used).

**Comment by supermerill** [comment]
On gyroid, it uses the z position and flow.spacing() to create the waves. I don't think there are something else.
Spacing change when the layer height change, because the half-circle is smaller. Maybe "width" should be used instead of "spacing"?

**Comment by ctmakro** [comment]
can confirm this bug. Adaptive causes infill pattern to shift abruptly.

**Comment by TD5023** [comment]
> Adaptive is probably breaking one of assumptions in that codebase, that or **the line thickness is changing (which would rescale the polygon used).**

I'll also confirm that the infill shifts on layer height changes, and that it's an issue across all infill patterns. Regarding line thickness variance, that's not the problem because the shift is still there with a constant thickness defined. Hopefully this is on someone's radar, because while it's not a huge deal for straight-line infills, I _really_ would like to be able to use gyroid and adaptive, but it's not feasible right now.

**Comment by lordofhyphens** [comment]
It's because, as of the last time looked into it, the % setting was based on the layer height for that layer.

**Comment by TD5023** [comment]
Yes, that's true if you use a % setting. However, it also comes with the option to set a constant thickness (e.g. 0.45 mm), so that should eliminate the need for spacial recalculations on non-solid infill. Even with a hard-defined thickness, the infill still shifts.

It should also be noted that all it does is shift. Even when a % is used, there is no change to the line density, which would be expected if the thickness were the issue.

**Comment by lordofhyphens** [comment]
When I said %, I was explicitly referring to infill_density.

**Comment by TD5023** [comment]
My mistake. I thought you were referring to line thickness calculations. I guess I always assumed infill density was based on a 2d snapshot of the layer, but if layer height is a factor and it's based on mass extruded, a slight change would make sense.
